src/include/storage/recovery/recovery_manager.h

```diff
@@ -0,0 +1,275 @@
+    /**
+     * @param recovery_manager pointer to recovery manager who initialized task
+     */
+    explicit RecoveryTask(RecoveryManager *recovery_manager) : recovery_manager_(recovery_manager) {}
+
+    /**
+     * Runs the recovery task. Our task only calls Recover on the log manager.
+     */
+    void RunTask() override { recovery_manager_->Recover(); }
+
+    /**
+     * Terminate does nothing, the task will terminate when RunTask() returns. In the future if we need to support
+     * interrupting recovery, this can be handled here.
+     */
+    void Terminate() override {}
+
+   private:
+    RecoveryManager *recovery_manager_;
+  };
+
+ public:
+  /**
+   * @param log_provider arbitrary provider to receive logs from
+   * @param catalog system catalog to interface with sql tables
+   * @param txn_manager txn manager to use for re-executing recovered transactions
+   * @param thread_registry thread registry to register tasks
+   * @param store block store used for SQLTable creation during recovery
+   */
+  explicit RecoveryManager(AbstractLogProvider *log_provider, common::ManagedPointer<catalog::Catalog> catalog,
+                           transaction::TransactionManager *txn_manager,
+                           common::ManagedPointer<terrier::common::DedicatedThreadRegistry> thread_registry,
+                           BlockStore *store)
+      : DedicatedThreadOwner(thread_registry),
+        log_provider_(log_provider),
+        catalog_(catalog),
+        txn_manager_(txn_manager),
+        block_store_(store),
+        recovered_txns_(0) {}
+
+  /**
+   * Starts a background recovery task. Recovery will fully recover until the log provider stops providing logs.
+   */
+  void StartRecovery() {
+    TERRIER_ASSERT(recovery_task_ == nullptr, "Recovery already started");
+    recovery_task_ =
+        thread_registry_->RegisterDedicatedThread<RecoveryTask>(this /* dedicated thread owner */, this /* task arg */);
+  }
+
+  /**
+   * Stops the background recovery task. This will block until recovery finishes, if it has not already.
+   */
+  void FinishRecovery() {
+    TERRIER_ASSERT(recovery_task_ != nullptr, "Recovery must already have been started");
+    bool result UNUSED_ATTRIBUTE =
+        thread_registry_->StopTask(this, recovery_task_.CastManagedPointerTo<common::DedicatedThreadTask>());
+    TERRIER_ASSERT(result, "Task termination should always succeed");
+  }
+
+  /**
+   * @return number of committed txns recovered so far
+   */
+  uint32_t GetRecoveredTxnCount() const { return recovered_txns_; }
+
+ private:
+  friend class RecoveryTests;
+
+  // Log provider for reading in logs
+  AbstractLogProvider *log_provider_;
+
+  // Catalog to fetch table pointers
+  common::ManagedPointer<catalog::Catalog> catalog_;
+
+  // Transaction manager to create transactions for recovery
+  transaction::TransactionManager *txn_manager_;
+
+  // Block store, used to create tables during recovery
+  BlockStore *block_store_;
```

Comment:
Right now, the execution engine and/or something else in the system determines which block stores get used for which table.  Somehow we should be allowing that same system to control backing block stores in recovery as well.

Unfortunately, I have no clue what that would look like since I'm not actually sure which component is making that determination.  It's possible that we need to implement `pg_tablespace` (appropriately modified to match in-memory semantics) and use that to store the information so that you are consistent in recovery.